OCPBUGS-38432: Added step to networking Converting HTTP header case s… (#80474)
openshift/openshift-docs: modules/nw-ingress-converting-http-header-case.adoc

[id="nw-ingress-converting-http-header-case_{context}"]
= Converting HTTP header case

HAProxy lowercases HTTP header names by default; for example, changing `Host: xyz.com` to `host: xyz.com`. If legacy applications are sensitive to the capitalization of HTTP header names, use the Ingress Controller `spec.httpHeaders.headerNameCaseAdjustments` API field for a solution to accommodate legacy applications until they can be fixed.

[IMPORTANT]
====

> {product-title} includes HAProxy 2.8. If you want to upgrade to this version of the web-based load balancer, ensure that add `spec.httpHeaders.headerNameCaseAdjustments` to your cluster's configuration file.

Reviewer: Update is the preferred term here:
```suggestion
{product-title} includes HAProxy 2.8. If you want to update to this version of the web-based load balancer, ensure that add `spec.httpHeaders.headerNameCaseAdjustments` to your cluster's configuration file.
```

Author: :heavy_check_mark:

> {product-title} includes HAProxy 2.8. If you want to upgrade to this version of the web-based load balancer, ensure that add `spec.httpHeaders.headerNameCaseAdjustments` to your cluster's configuration file.

Reviewer: I think the phrase "ensure that add `spec.httpHeaders.headerNameCaseAdjustments` to your cluster's configuration file." is missing something or needs to be reworked, could you take a look and readjust depending on the intended meaning?

Author: Updated to state:

> ensure that you add the `spec.httpHeaders.headerNameCaseAdjustments` section to your cluster's configuration file.

====

As a cluster administrator, you can convert the HTTP header case by entering the `oc patch` command or by setting the `HeaderNameCaseAdjustments` field in the Ingress Controller YAML file.

.Prerequisites

* You have installed the OpenShift CLI (`oc`).
* You have access to the cluster as a user with the `cluster-admin` role.

.Procedure

* Capitalize an HTTP header by using the `oc patch` command.

> . Enter the `oc patch` command to change the HTTP header from `host` to `Host`:

Reviewer: I think that if this is meant to be a substep, this should use a level 2 ordered list item, even if the parent list item is unordered. I don't know if there's a style guide ruling on this but it would just seem a bit confusing imo to use level one items in this context:
```suggestion
.. Enter the `oc patch` command to change the HTTP header from `host` to `Host`:
```

If for some reason the formatting doesn't work when you change it, then this can be left as is.

Author: :heavy_check_mark:

> . Enter the `oc patch` command to change the HTTP header from `host` to `Host`:

Reviewer: This is out of scope to your changes so feel free to ignore this item, but I think this sort of format would better align with the way our [docs guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#procedure-step-that-introduces-a-command) suggest we introduce commands:
```suggestion
. Change the HTTP header from `host` to `Host` by running the following command:
```

Author: :heavy_check_mark:

+
[source,terminal]
----
$ oc -n openshift-ingress-operator patch ingresscontrollers/default --type=merge --patch='{"spec":{"httpHeaders":{"headerNameCaseAdjustments":["Host"]}}}'
----
+

> . Create a `Route` resource YAML file so that the annotation can be applied to the application.

Reviewer: Similar comment about making this a level 2 list item:
```suggestion
.. Create a `Route` resource YAML file so that the annotation can be applied to the application.
```

Author: :heavy_check_mark:

+
.Example of a route named `my-application`
[source,yaml]
----
apiVersion: route.openshift.io/v1
kind: Route
metadata:
  annotations:
    haproxy.router.openshift.io/h1-adjust-case: true <1>
  name: <application_name>
  namespace: <application_name>
# ...
----

> <1> Set `haproxy.router.openshift.io/h1-adjust-case` so that Ingress Controller can adjust the `host` request header as specified. 

Reviewer: Should there be an article here?
```suggestion
<1> Set `haproxy.router.openshift.io/h1-adjust-case` so that the Ingress Controller can adjust the `host` request header as specified. 
```

* Specify adjustments by configuring the `HeaderNameCaseAdjustments` field in the Ingress Controller YAML configuration file.

> . The following example Ingress Controller YAML file adjusts the `host` header to `Host` for HTTP/1 requests to appropriately annotated routes:

Reviewer: Same comment as above:
```suggestion
.. The following example Ingress Controller YAML file adjusts the `host` header to `Host` for HTTP/1 requests to appropriately annotated routes:
```

Author: :heavy_check_mark:

$ oc -n openshift-ingress-operator patch ingresscontrollers/default --type=merge --patch='{"spec":{"httpHeaders":{"headerNameCaseAdjustments":["Host"]}}}'
----
+

> . Create a `Route` resource YAML file.

Reviewer 2: ```suggestion
. Create a `Route` resource YAML file so that the annotation can be applied to the pod.
```

Author: :heavy_check_mark:

+
.Example of a route named `my-application`
[source,yaml]
----
apiVersion: route.openshift.io/v1
kind: Route
metadata:
 name: my-application

>  namespace: default

Reviewer 3: namespace: my-application
@dfitzmau  Kindly change the namespace from default to  my-application

Author: Thanks, Melvin. Updated.

spec:
 to:
  kind: Service

>   name: my-application

Reviewer 4: I'm wondering about the difference of this `Route` object and the one down the procedure: 1) is it on purpose that they are from different namespaces? 2) is there a purpose in adding the annotation in 2 different ways (`oc annotate` in the yaml directly)?
I think I would prefer to see similar yamls for both to make the user experience smoother.

Reviewer 3: i agree to Andrey, 
1. it should be in same name space
2. this was already created long back. i guess these are two different methods to configure. one by patching the controller and annotating the route and another by changing directly in the controller and route yaml files.

Even after the above both steps, the yaml files of the controller and route will be same for both methods.

Author: Thanks, Andrey and Melvin. I updated the namespace name.

    - Host
----
+

> . The following example route enables HTTP response header name case adjustments by using the `haproxy.router.openshift.io/h1-adjust-case` annotation:

Reviewer: ```suggestion
.. The following example route enables HTTP response header name case adjustments by using the `haproxy.router.openshift.io/h1-adjust-case` annotation:
```

Author: :heavy_check_mark:
